### PR TITLE
Allow `Raster`s as inputs for `Coreg.fit()` and `Coreg.apply()`

### PR DESCRIPTION
**Pull request by erikmannerfelt**
Solves #174.

See the issue for the new syntax!

### CODE REVIEW

File: tests/test_coreg.py [hunk 1]
```diff
@@ -484,6 +485,57 @@ def test_blockwise_coreg(self, pipeline, subdivision):
         # Check that offsets were actually calculated.
         assert np.sum(np.abs(np.linalg.norm(stats[["x_off", "y_off", "z_off"]], axis=0))) > 0
 
+    def test_coreg_raster_and_ndarray_args(_) -> None:
```
Comment (rhugonnet) [inline tests/test_coreg.py:488]: Is `_` playing a role in the test function arguments? I've only ever seen `self`

File: xdem/coreg.py [hunk 2]
```diff
@@ -468,6 +469,20 @@ def fit(self, reference_dem: Union[np.ndarray, np.ma.masked_array],
         if weights is not None:
             raise NotImplementedError("Weights have not yet been implemented")
 
+        # If both DEMs are Rasters, validate that 'dem_to_be_aligned' is in the right grid. Then extract its data.
+        if isinstance(dem_to_be_aligned, gu.Raster) and isinstance(reference_dem, gu.Raster):
+            dem_to_be_aligned = dem_to_be_aligned.reproject(reference_dem, silent=True).data
```
Comment (rhugonnet) [inline xdem/coreg.py:474]: after seeing so many `silent=True`, I'm wondering if we'd not prefer `silent=True` by default on `Raster.reproject` as well!

File: xdem/coreg.py [hunk 3]
```diff
@@ -511,8 +526,18 @@ def fit(self, reference_dem: Union[np.ndarray, np.ma.masked_array],
         # Flag that the fitting function has been called.
         self._fit_called = True
 
-    def apply(self, dem: Union[np.ndarray, np.ma.masked_array],
-              transform: rio.transform.Affine) -> Union[np.ndarray, np.ma.masked_array]:
+    @overload
```
Comment (rhugonnet) [inline xdem/coreg.py:529]: very nice

File: xdem/coreg.py [hunk 4]
```diff
@@ -468,6 +469,20 @@ def fit(self, reference_dem: Union[np.ndarray, np.ma.masked_array],
...
+        if isinstance(dem_to_be_aligned, gu.Raster) and isinstance(reference_dem, gu.Raster):
+            dem_to_be_aligned = dem_to_be_aligned.reproject(reference_dem, silent=True).data
+        # If 'reference_dem' is not a Raster, just extract the data (shape validation comes later)
+        elif isinstance(dem_to_be_aligned, gu.Raster):
+            dem_to_be_aligned = dem_to_be_aligned.data
+
+        # If 'reference_dem' is a Raster, use its transform (if not given) and then extract its data.
+        if isinstance(reference_dem, gu.Raster):
+
+            if transform is None:
```
Comment (rhugonnet) [inline xdem/coreg.py:482]: I wonder if maybe we should raise a RunTimeWarning if a `RasterType` is provided but another `transform` is given and used? 
That seems quite dangerous to me, and the user should be informed! 
And, in practice, it won't happen so often that the Warning will become annoying. It's really rare to provide a different transform than that of your Raster

File: tests/test_coreg.py [hunk 5]
```diff
@@ -379,7 +380,7 @@ def test_z_scale_corr(self):
         zcorr.fit(self.ref.data, scaled_dem)
 
         # Apply the correction
-        unscaled_dem = zcorr.apply(scaled_dem, None)
+        unscaled_dem = zcorr.apply(scaled_dem, self.ref.transform)
```
Comment (rhugonnet) [inline tests/test_coreg.py:383]: I'm not sure I get why those needed changing?

**Review by rhugonnet — APPROVED** [review]
Tip top!
Birds and heat have a much more productively-positive effect on you than me :laughing:

File: xdem/coreg.py (hunk 2 above)
Comment (erikmannerfelt) [reply xdem/coreg.py:474]: A difference is that `Raster.reproject(..., silent=True)` may actually have some unforeseen consequences: it returns a copy, so any action after that will propagate to the original raster too.

While I'm not perfectly happy with the warning that shows up here and there, it would almost have to be replaced by another protection mechanism to be completely safe....

File: tests/test_coreg.py (hunk 5 above)
Comment (erikmannerfelt) [reply tests/test_coreg.py:383]: `Raster.apply()` now raises an error if an array is provided but no transform. This makes the tools perfectly consistent (there was no previous mechanism to validate that a transform was given, so the error messages were very confusing). In this particular case, it's a bit redundant, but I think that's an okay sacrifice for consistency! If you and @adehecq disagree, I'm sure we can find another way.

File: xdem/coreg.py (hunk 2 above)
Comment (rhugonnet) [reply xdem/coreg.py:474]: True, and it's not everyday we reproject on same CRS.
Good as is!

**Comment by rhugonnet** [comment]
Actually, a big question raised by this:
With https://github.com/GlacioHack/GeoUtils/pull/210 and this PR, we now have many use cases where we can use `np.ndarray` and `Raster` objects interchangeably. But there is still several functions for which it does not work, right? Such as terrain attributes.

Should we aim to have xdem working with np.ndarrays and Raster objects interchangeably? If yes, what is left to do? (I think that, @erikmannerfelt, you have the best viewpoint on all this!)
Also, it should be crystal clear when a Raster can be provided simply as a **convenience** and a np.ndarray would suffice, or when georeferenced data is **absolutely necessary**. It will get blurry for the users if we allow to provide both everywhere for convenience, so I'm not sure how to tackle this!

**Comment by erikmannerfelt** [comment]
> Actually, a big question raised by this:
> With [GlacioHack/GeoUtils#210](https://github.com/GlacioHack/GeoUtils/pull/210) and this PR, we now have many use cases where we can use `np.ndarray` and `Raster` objects interchangeably. But there is still several functions for which it does not work, right? Such as terrain attributes.
> 
> Should we aim to have xdem working with np.ndarrays and Raster objects interchangeably? If yes, what is left to do? (I think that, @erikmannerfelt, you have the best viewpoint on all this!)
> Also, it should be crystal clear when a Raster can be provided simply as a **convenience** and a np.ndarray would suffice, or when georeferenced data is **absolutely necessary**. It will get blurry for the users if we allow to provide both everywhere for convenience, so I'm not sure how to tackle this!

Since, `geoutils==0.0.3`, `Raster`s implement the array interface, so they _should_ be possible to use interchangeably with ndarrays. The tests don't cover that "_should_" part, however, and `Raster`s are not consistently returned if provided as an argument. For now, all functions that show a `Raster`/`RasterType` argument/return type hint returns a `RasterType` if given a `RasterType`, and those who don't will return an `ndarray`/`masked_array`.

Long story short, the only way to validate how interchangeable they are is to include more tests!!

File: xdem/coreg.py (hunk 4 above)
Comment (erikmannerfelt) [reply xdem/coreg.py]: Added in cc0cdb3!

I accidentally included it with a merge... Jaja, I blame the 3 AM birds!

File: tests/test_coreg.py (hunk 1 above)
Comment (erikmannerfelt) [reply tests/test_coreg.py:488]: An argument needs to be provided as it is a method, but this particular method does not actually need `self`. Potentially, this should be its own function, but currently all `Coreg`-related tests are in the `TestCoregClass`, so it felt wrong to put it outside by itself!

**Comment by erikmannerfelt** [comment]
I think I've fixed all of @rhugonnet 's comments.

Any other thoughts, @rhugonnet  and @adehecq ?

**Review by rhugonnet — APPROVED** [review]
All good for me! 
Maybe we should open an issue on the "interchangeable" discussion? 
It would be good if we could say "All xdem functions accept interchangeably `np.ndarray` and `Raster` subclass objects"

**Comment by erikmannerfelt** [comment]
Already done in #178 !

Thanks, @rhugonnet !

File: tests/test_coreg.py (hunk 5 above)
Comment (adehecq) [reply tests/test_coreg.py:383]: Sounds ok to me

**Comment by adehecq** [comment]
Fantastic !
